nemo/core/optim/lr_scheduler.py

```diff
@@ -0,0 +1,412 @@
+
+    def _get_lr(self, step):
+        new_lrs = [
+            _poly_decay(
+                initial_lr,
+                step=step - self.hold_steps,
+                decay_steps=self.max_steps - max(self.warmup_steps, self.hold_steps),
+                power=self.power,
+                min_lr=self.min_lr,
+                cycle=self.cycle,
+            )
+            for initial_lr in self.base_lrs
+        ]
+        return new_lrs
+
+
+def prepare_lr_scheduler(
+    optimizer: optim.Optimizer,
+    scheduler_config: Dict[str, Any],
+    train_dataloader: Optional[dataloader.DataLoader] = None,
+) -> Optional[Dict[str, Any]]:
+    """
+    Constructs an LR Scheduler (optionally) for a given optimizer, based on a config with the following schema
+
+    {
+    ...,
+    "scheduler": <a class that inherits torch.optim.lr_scheduler._LRScheduler>,
+    "scheduler_args": {
+        "max_steps": int, <OR> "iters_per_batch": int,
+        "monitor": <metric to monitor; say "loss" or "val_loss">,
+        <any kwarg to pass onto the optimizer>
+      }
+    }
+
+    Args:
+        optimizer: An instantiated Optimizer.
+        scheduler_config: A dictionary which follows the above schema.
+        train_dataloader: Optional requirement, must be passed if "iters_per_batch" is defined
+            instead of "max_steps". Used to compute effective "max_steps".
+
+    Returns:
+        A dictionary containing the LR Scheduler implementation if the config was successfully parsed
+        along with other parameters required by Pytorch Lightning, otherwise None.
+    """
+    if 'scheduler' in scheduler_config:
+        if 'scheduler_args' in scheduler_config:
+            scheduler_args = scheduler_config['scheduler_args']
+        else:
+            raise ValueError("If `scheduler` is provided, `scheduler_args` must be provided.")
+
+    else:
+        logging.info('Scheduler not initialized as no `scheduler` argument supplied to setup_optimizer()')
+        return None
+
+    # Get the scheduler class from the config
+    scheduler = scheduler_config['scheduler']
+
+    # Extract value to monitor in losses, if provided.
+    if 'monitor' in scheduler_args:
+        monitor = scheduler_args.pop('monitor')
+    else:
+        # default to train loss
+        monitor = 'loss'
+
+    # Compute effective max_steps if iters_per_batch is provided
+    if 'iters_per_batch' in scheduler_args:
+        if train_dataloader is None:
+            raise ValueError(
+                'As `iters_per_batch` is provided, it is required to pass the train dataloader in order '
+                'to compute effective maximum number of steps'
+            )
+
+        iters_per_batch = scheduler_args.pop('iters_per_batch')
+        num_samples = len(train_dataloader.dataset)
+        batch_size = train_dataloader.batch_size
+        max_steps = round(num_samples * iters_per_batch / float(batch_size))
```

Comment:
Can you explain this part to me?

Shouldn't it be
```
steps_in_epoch = len(train_data_layer) / (args.batch_size * args.iter_per_step * nf.world_size)
max_steps = steps_in_epoch * num_epochs
```

Reply:
We don't have the notion of drop last epoch right now, though we can add it in the future. Round() is basically the switch between ceil and floor for us without the flag for the moment. 

Iters per step is not accumulate grad batches, nf.world size is gpus * num_nodes. Right now I am expecting user to precompute the epochs / denominator (disregarding batch size in denom and number of samples in numerator) and then I inject those values from the train data layer. Once we are capable of passing config objects, we can hide this internal computation.